src/complete.rs

```diff
@@ -88,28 +166,63 @@ pub fn lf<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, char,
 ///         ;  headers and use with caution in
 ///         ;  other contexts.
 // code as equivalent avoid branching LWSP = *([CRLF] WSP)
-pub fn lwsp<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, &str, E> {
+pub fn lwsp<I, E>(input: I) -> IResult<I, I, E>
+where
+    I: Clone
+        + Offset
+        + PartialEq
+        + InputIter
+        + Slice<RangeTo<usize>>
+        + Slice<RangeFrom<usize>>
+        + InputLength,
+    <I as InputIter>::Item: AsChar,
+    E: ParseError<I>,
+{
     recognize(many0_count(terminated(opt(crlf), wsp)))(input)
 }
 
 /// OCTET = %x00-FF ; 8 bits of data
-pub fn octet(input: &[u8]) -> IResult<&[u8], &[u8]> {
-    take(1usize)(input)
+pub fn octet<E>(input: &[u8]) -> IResult<&[u8], u8, E>
+where
+    for<'a> E: ParseError<&'a [u8]>,
+{
+    match input.split_first() {
+        None => Err(OutCome::Error(E::from_error_kind(
+            input,
+            ErrorKind::Complete,
+        ))),
+        Some((&b, tail)) => Ok((tail, b)),
+    }
```

Comment:
unfortunately I didn't find a good way in my opinion to implement this. I limited this to `&[u8]` only and I believe it's make sense, for example I pretty sure NUL is invalid value for a char in rust.

Avoiding generic here allow to be O(1) (assuming optimization is not smart enough) for bound checking and I don't like unwrap thing.

I choice `ErrorKind::Complete` cause that was the only one that make sense here. I believe it's quite good. I didn't want to create a custom error just for this.